lib/Mojolicious/Guides/Tutorial.pod

```diff
@@ -0,0 +1,938 @@
+L<Mojo::Transaction::WebSocket/"json"> with L<Mojolicious::Controller/"on"> and return them with
+L<Mojolicious::Controller/"send">.
+
+  use Mojolicious::Lite -signatures;
+
+  websocket '/echo' => sub ($c) {
+    $c->on(json => sub ($c, $hash) {
+      $hash->{msg} = "echo: $hash->{msg}";
+      $c->send({json => $hash});
+    });
+  };
+
+  get '/' => 'index';
+
+  app->start;
+  __DATA__
+
+  @@ index.html.ep
+  <!DOCTYPE html>
+  <html>
+    <head>
+      <title>Echo</title>
+      <script>
+        const ws = new WebSocket('<%= url_for('echo')->to_abs %>');
+        ws.onmessage = function (event) {
+          document.body.innerHTML += JSON.parse(event.data).msg;
+        };
+        ws.onopen = function (event) {
+          ws.send(JSON.stringify({msg: 'I ♥ Mojolicious!'}));
+        };
+      </script>
+    </head>
+  </html>
+
+For more information about real-time web features see also L<Mojolicious::Guides::Cookbook/"REAL-TIME WEB">.
+
+=head2 Mode
+
+You can use the L<Mojo::Log> object from L<Mojolicious/"log"> to portably collect debug messages and automatically
+disable them later in a production setup by changing the L<Mojolicious> operating mode, which can also be retrieved
+from the attribute L<Mojolicious/"mode">.
+
+  use Mojolicious::Lite -signatures;
+
+  # Prepare mode specific message during startup
+  my $msg = app->mode eq 'development' ? 'Development!' : 'Something else!';
+
+  get '/' => sub ($c) {
+    $c->app->log->debug('Rendering mode specific message');
+    $c->render(text => $msg);
+  };
+
+  app->log->debug('Starting application');
+  app->start;
+
+The default operating mode will usually be C<development> and can be changed with command line options or the
+C<MOJO_MODE> and C<PLACK_ENV> environment variables. A mode other than C<development> will raise the log level from
+C<trace> to C<info>. All messages will be written to C<STDERR> by default.
+
+  $ ./myapp.pl daemon -m production
+
+Mode changes also affect a few other aspects of the framework, such as the built-in C<exception> and C<not_found>
+pages. Once you switch modes from C<development> to C<production>, no sensitive information will be revealed on those
+pages anymore.
+
+=head2 Testing
+
+Testing your application is as easy as creating a C<t> directory and filling it with normal Perl tests like
+C<t/basic.t>, which can be a lot of fun thanks to L<Test::Mojo>.
+
+  use Test::More;
+  use Mojo::File qw(curfile);
+  use Test::Mojo;
+
+  # Portably point to "../myapp.pl"
+  my $script = curfile->dirname->sibling('myapp.pl');
+
+  my $t = Test::Mojo->new($script);
+  $t->get_ok('/')->status_is(200)->content_like(qr/Funky/);
+
+  done_testing();
+
+Just run your tests with L<prove>.
+
+  $ prove -l -v
+  $ prove -l -v t/basic.t
+
+=head1 MORE
+
+You can continue with L<Mojolicious::Guides> now or take a look at the L<Mojolicious
+wiki|https://github.com/mojolicious/mojo/wiki>, which contains a lot more documentation and examples by many different
```

Comment:
While there is certainly some good info in the wiki, it hasn't been updated in 3 years!  D:  I wouldn't want to not send people there, but also the 3+yo staleness of the information could really turn people off!